core/commonMain/src/ArgParser.kt

```diff
@@ -342,24 +371,117 @@ open class ArgParser(
+        }
+        return false
+    }
 
     /**
-     * Try to recognize command line element as short form of option.
+     * Save option without parameter.
+     *
+     * @param argValue argument value with all information about option.
+     */
+    private fun saveOptionWithoutParameter(argValue: ParsingValue<*, *>) {
+        // Boolean flags.
+        if (argValue.descriptor.fullName == "help") {
+            println(makeUsage())
+            exitProcess(0)
+        }
+        saveAsOption(argValue, "true")
+    }
+
+    /**
+     * Save option described with standard separated form `--name value`.
+     *
+     * @param argValue argument value with all information about option.
+     * @param argIterator iterator over command line arguments.
+     */
+    private fun saveStandardOptionForm(argValue: ParsingValue<*, *>, argIterator: Iterator<String>) {
+        if (argValue.descriptor.type.hasParameter) {
+            if (argIterator.hasNext()) {
+                saveAsOption(argValue, argIterator.next())
+            } else {
+                // An error, option with value without value.
+                printError("No value for ${argValue.descriptor.textDescription}")
+            }
+        } else {
+            saveOptionWithoutParameter(argValue)
+        }
+    }
+
+    /**
+     * Try to recognize and save command line element as short form of option.
      *
      * @param candidate string with candidate in options.
+     * @param argIterator iterator over command line arguments.
      */
-    private fun recognizeOptionShortForm(candidate: String) =
-            if (candidate.startsWith(optionShortFromPrefix))
-                shortNames[candidate.substring(optionShortFromPrefix.length)]
-            else null
+    private fun recognizeAndSaveOptionShortForm(candidate: String, argIterator: Iterator<String>): Boolean {
+        if (!candidate.startsWith(optionShortFromPrefix)) return false
+        // Try to find exact match.
+        val option = candidate.substring(optionShortFromPrefix.length)
+        val argValue = shortNames[option]
+        if (argValue != null) {
+            saveStandardOptionForm(argValue, argIterator)
+        } else {
+            if (prefixStyle != OptionPrefixStyle.GNU)
+                return false
+
+            // Try to find collapsed form.
+            val firstOption = shortNames["${option[0]}"] ?: return false
+            // Form with value after short form without separator.
+            if (firstOption.descriptor.type.hasParameter) {
+                saveAsOption(firstOption, option.substring(1))
+            } else {
+                // Form with several short forms as one string.
+                val optionsList = mutableListOf(firstOption)
+                val otherBooleanOptions = option.substring(1)
+                for (option in otherBooleanOptions) {
+                    if (shortNames["$option"]?.descriptor?.type?.hasParameter != false) {
+                        return false
```

Comment:
So `-abc` can only be treated as several options if all of them are booleans?
If `a` has parameter then `bc` is its argument, and if `a` doesn't, but any of `b` or `c` does, then it's an invalid option, right? In that case I'd expect to error immediately instead of exiting with false and handling this arg as not an option.

Reply:
But it still can be an argument if there are expected arguments. We can't predict the argument format.

Reply:
How can it be an argument, if it begin with `-`?

Reply:
It can be a negative number at least. And there are other options in different cases. I haven't seen limitation in other cli parsers that argument can't start with `-`.

Reply:
Still, if `a` is an option, I wouldn't expect `-abc` parsed as an argument in any case (for example `c` is not a boolean option).

Reply:
I don't understand the motivation to provide such limitations. I saw the benchmarking command line tools that took as arguments extra options of compiler/linker. But for GNU options ok, we can do so, because there is an opportunity to provide such arguments after `--`

Reply:
My motivation behind this is to avoid confusing behavior when user has mistyped option keys, which is not uncommon when using the command line interface. At best, they get some unhelpful message, like "argument -abv was not expected", at worst — that option is gonna be interpreted as an argument.

Reply:
Ok, I have already fixed this.